File: GFG POTD/011-October-2024/023-find-the-sum-of-last-n-nodes-of-the-linked-list.cpp
// https://www.geeksforgeeks.org/problems/find-the-sum-of-last-n-nodes-of-the-linked-list/1



//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node *next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};

// } Driver Code Ends
class Solution {
  public:
    /*Structure of the node of the linled list is as
    struct Node {
        int data;
        struct Node* next;

        Node(int x){
            data = x;
            next = NULL;
        }
    };
    */
    
    // TC : O(N)
    // SC : O(N) Recursive Depth
    Node* reverse(Node* head) {
        if (!head || !head -> next) 
            return head;
        
        Node* newH = reverse(head -> next);
        head -> next -> next = head;
        head -> next = NULL;
        
        return newH;
    }
    
    // TC : O(N)
    // SC : O(N) recursive call stack
    void solve(Node* head, int n, int &result, int &cnt) {
        if (head == NULL) return;

        solve(head -> next, n, result, cnt);
        cnt += 1;
        if (cnt <= n) result += head -> data;
    }
    
    // TC : O(N)
    // SC : O(N)
    int approach_1(struct Node* head, int n) {
        Node* reversedHead = reverse(head);
        int sum = 0;
        int k = n;
        Node* temp = reversedHead;
        while (k -- && temp) {
            sum += temp -> data;
            temp = temp -> next;
        }
        return sum;
    }
    
    // TC : O(N)
    // SC : O(1)
    int approach_2(struct Node* head, int n) {
        int result = 0;
        int cnt = 0;
        solve(head, n, result, cnt);
        return result;
    }
    
    // your task is to complete this function
    // function should return sum of last n nodes
    int sumOfLastN_Nodes(struct Node* head, int n) {
        // return approach_1(head, n);  // Reverse and sum till n nodes
        return approach_2(head, n);  // Recursive
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int n;
        cin >> n;
        cin.ignore();
        struct Node *head = new Node(arr[0]);
        struct Node *tail = head;
        for (int i = 1; i < arr.size(); ++i) {
            tail->next = new Node(arr[i]);
            tail = tail->next;
        }
        Solution ob;
        cout << ob.sumOfLastN_Nodes(head, n) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: GFG POTD/011-October-2024/023-find-the-sum-of-last-n-nodes-of-the-linked-list.java
// https://www.geeksforgeeks.org/problems/find-the-sum-of-last-n-nodes-of-the-linked-list/1




//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int a) {
        data = a;
        next = null;
    }
}
// } Driver Code Ends
/*
Node class is as follows:

class Node {
    int data;
    Node next;

    public Node (int data){
        this.data = data;
        this.next = null;
    }
}
*/

class Solution {
    // Return the sum of last k nodes
    public int sumOfLastN_Nodes(Node head, int n) {
        int result[] = {0};
        int cnt[] = {0};
        solve(head, n, result, cnt);
        return result[0];
    }
    
    void solve(Node head, int n, int result[], int cnt[]) {
        if (head == null) return;
        
        solve(head.next, n, result, cnt);
        cnt[0] += 1;
        if (cnt[0] > n) return;
        result[0] += head.data;
    }
}


//{ Driver Code Starts.
class LinkedList {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            ArrayList<Integer> arr = new ArrayList<>();
            String input = br.readLine();
            StringTokenizer st = new StringTokenizer(input);
            while (st.hasMoreTokens()) {
                arr.add(Integer.parseInt(st.nextToken()));
            }
            int n = Integer.parseInt(br.readLine());
            Node head = new Node(arr.get(0));
            Node tail = head;
            for (int i = 1; i < arr.size(); ++i) {
                tail.next = new Node(arr.get(i));
                tail = tail.next;
            }
            Solution ob = new Solution();
            System.out.println(ob.sumOfLastN_Nodes(head, n));
        }
    }
}
// } Driver Code Ends